packages/engine/paima-funnel/src/funnels/carp/funnel.ts

```diff
@@ -0,0 +1,269 @@
+    // recoverState
+    const lastTimestamp = await timeout(
+      this.sharedData.web3.eth.getBlock(blockHeight - 1),
+      DEFAULT_FUNNEL_TIMEOUT
+    );
+
+    let grouped = await readDataInternal(
+      this.bufferedData,
+      this.carpUrl,
+      this.sharedData.extensions,
+      lastTimestamp.timestamp as number
+    );
+
+    const composed = composeChainData(this.bufferedData, grouped);
+
+    this.bufferedData = null;
+
+    return composed;
+  }
+
+  public override async readPresyncData(
+    args: ReadPresyncDataFrom
+  ): Promise<{ [network: number]: PresyncChainData[] | 'finished' }> {
+    const arg = args.find(arg => arg.network == Network.CARDANO);
+
+    let basePromise = this.baseFunnel.readPresyncData(args);
+
+    if (arg && arg.from >= 0 && arg.from < this.startSlot) {
+      const [poolEvents, data] = await Promise.all([
+        Promise.all(
+          this.sharedData.extensions
+            .filter(extension => extension.cdeType === ChainDataExtensionType.CardanoPool)
+            .map(extension => {
+              const data = getCdePoolData(
+                this.carpUrl,
+                extension as ChainDataExtensionCardanoDelegation,
+                arg.from,
+                Math.min(arg.to, this.startSlot - 1),
+                slot => {
+                  return slot;
+                }
+              );
+              return data;
+            })
+        ),
+        basePromise,
+      ]);
+
+      let grouped = groupCdeData(Network.CARDANO, arg.from, arg.to, poolEvents);
+
+      if (grouped.length > 0) {
+        data[Network.CARDANO] = grouped;
+      }
+
+      return data;
+    } else {
+      const data = await basePromise;
+
+      if (arg) {
+        data[Network.CARDANO] = 'finished';
+      }
+
+      return data;
+    }
+  }
+
+  public static async recoverState(
+    sharedData: FunnelSharedData,
+    dbTx: PoolClient,
+    baseFunnel: ChainFunnel,
+    carpUrl: string,
+    startingBlockHeight: number
+  ): Promise<CarpFunnel> {
+    const startingSlot = timestampToAbsoluteSlot(
+      (await sharedData.web3.eth.getBlock(startingBlockHeight)).timestamp as number
+    );
+
+    return new CarpFunnel(sharedData, dbTx, baseFunnel, carpUrl, startingSlot);
+  }
+}
+
+async function readDataInternal(
+  data: ChainData[],
+  carpUrl: string,
+  extensions: ChainDataExtension[],
+  lastTimestamp: number
+): Promise<PresyncChainData[]> {
+  // the lower range is exclusive
+  const min = timestampToAbsoluteSlot(lastTimestamp);
+  // the upper range is inclusive
+  const max = timestampToAbsoluteSlot(Math.max(...data.map(data => data.timestamp)));
+
+  while (true) {
+    const stableBlock = await timeout(
+      query(carpUrl, Routes.blockLatest, {
+        offset: Number(confirmationDepth),
+      }),
+      DEFAULT_FUNNEL_TIMEOUT
+    );
+
+    if (stableBlock.block.slot > max) {
+      break;
+    }
+
+    await delay(1000);
+  }
```

Comment:
It can apply to syncing, but also a block may not have been produced yet. 

This is here in case the block at depth `confirmationDepth - 1` needs to be included in the range according to its timestamp/slot, but until an extra block is added at the tip then we don't want to consider it. This is because confirmation depends in depth, which is not necessarily related to time/slots.

Not sure how much it will matter in practice though, since it depends on the speed of both chains, but just in case.

Also, I guess the `1000` could be a constant.

Reply:
Okay, we should have a comment to explain this